grpcclient/apikey_setter_interceptor.go

```diff
@@ -25,9 +25,16 @@ import (
 
 func (c *LcClient) ApiKeySetterInterceptor() func(context.Context, string, interface{}, interface{}, *grpc.ClientConn, grpc.UnaryInvoker, ...grpc.CallOption) error {
 	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
-		if c.ApiKey != "" {
+		md, ok := metadata.FromOutgoingContext(ctx)
+		if ok {
+			authHeader, ok := md["lc-api-key"]
+			if ok && len(authHeader) > 0 && authHeader[0] != "" {
+				c.SetApiKey(authHeader[0])
```

Comment:
This looks strange to me.

The key could be taken from two different sources - either the LcClient instance (GetApiKey()) or the `ctx`. Such approach does not look clean to me. What's the use case here?

Additionally, I think setting the key from `ctx` in the LcClient object is very misleading and in fact error prone. `ctx` is supposed to hold the state for this particular request. Once the request is gone, we should not rely on the data retrieved from `ctx`. However here we store the API key in `LcClient` instance.

Consider the following scenario - if this interceptor can handle multiple simultaneous calls at the same time, each done with different API key then, assuming some bad timing, you can end up using someone else's key for authentication - that's because all those interceptors will concurrently set the key - so if interceptor of your request was already called but then some other interceptor got a chance to run before you retrieve the key, the API key may already be swapped.


To make this safe, this code should do the opposite - when context doesn't have the key, set it in this interceptor from `c.ApiKey` - and in other parts of the code, extract the API key from the context.